packages/artifact/src/internal-artifact-client.ts

```diff
@@ -100,25 +133,114 @@ export class DefaultArtifactClient implements ArtifactClient {
-    ): Promise<DownloadResponse> {
+      // Check only for downloadable items that are not directories. During upload, empty directories are rejected by the remote server so there should be no
+      // artifacts that consist of only empty directories
+      if (items.value.length === 0) {
+        throw new Error(
+          `No downloadable files were found for the artifact: ${artifactToDownload.name}`
+        )
+      } else {
+        const downloadSpecification = getDownloadSpecification(
+          name,
+          items.value,
+          path,
+          options?.createArtifactFolder || false
+        )
+        await createDirectoriesForArtifact(
+          downloadSpecification.directoryStructure
+        )
+        await downloadSingleArtifact(downloadSpecification.filesToDownload)
 
-    TODO
+        return {
+          artifactName: name,
+          downloadPath: downloadSpecification.rootDownloadLocation
+        }
+      }
+    }
   }
 
-  Downloads all artifacts associated with a run. Because there are multiple artifacts being downloaded, a folder will be created for each one in the specified or default directory
+  async downloadAllArtifacts(
+    path?: string | undefined
+  ): Promise<DownloadResponse[]> {
+    const response: DownloadResponse[] = []
+    const artifacts = await listArtifacts()
+    if (artifacts.count === 0) {
+      throw new Error(
+        `Unable to find any artifacts for the associated workflow`
+      )
+    }
 
-  export async function downloadAllArtifacts(
-      path?: string
-    ): Promise<DownloadResponse[]>{
+    if (!path) {
+      path = getWorkSpaceDirectory()
+    }
+    path = normalize(path)
+    path = resolve(path)
 
-      TODO
+    // Get container items for all artifacts. The resource url is the same for all artifacts so one call is sufficient
+    const items = await getContainerItemsForAllArtifacts(
+      artifacts.value[0].fileContainerResourceUrl
+    )
+    if (items.count === 0) {
+      core.warning(`No downloadable files were found for any artifacts`)
+    } else {
+      const ARTIFACT_CONCURRENCY = getDownloadArtifactConcurrency()
+      const parallelDownloads = [...new Array(ARTIFACT_CONCURRENCY).keys()]
+      let downloadedArtifacts = 0
+      await Promise.all(
+        parallelDownloads.map(async () => {
+          while (downloadedArtifacts < artifacts.count) {
+            const currentArtifactToUpload = artifacts.value[downloadedArtifacts]
+            downloadedArtifacts += 1
+
+            // Promise.All is not correctly inferring that 'path' is no longer possibly undefined: https://github.com/microsoft/TypeScript/issues/34925
+            const downloadSpecification = getDownloadSpecification(
+              currentArtifactToUpload.name,
+              items.value,
+              path!, // eslint-disable-line @typescript-eslint/no-non-null-assertion
+              true
+            )
+            await createDirectoriesForArtifact(
```

Comment:
This will not work cross platformly because the backing end is case insensitive.   We need to create folders as we download each individual files.

Reply:
The list of directories now comes from each individual file, I switched over to using `path.dirname` if the `itemType` is a file. Had to modify some tests and after some local testing it still works 😀 

This change has the side effect of not making every sub-directory be included in the `directoryStructure` result, but that's okay because everything gets created recursively

Example, before for the files `root/file1.txt` and `root/dir1/dir2/file2.txt` we would get 3 directories `root/ root/dir1/, root/dir1/dir2/`, now we only get 2: `root` and `root/dir1/dir2`

The directories still get created before the individual download starts (this allows for easier testing)